maven-core/src/main/java/org/apache/maven/graph/DefaultGraphBuilder.java

```diff
@@ -242,106 +260,86 @@ public DefaultGraphBuilder( BuildResumptionDataRepository buildResumptionDataRep
-                throw new MavenExecutionException( message, request.getPom() );
-            }
+        if ( excludedProjects.size() >= projectsToActivate.size() )
+        {
+            boolean isPlural = excludedProjects.size() > 1;
+            String message = String.format( "The project exclusion%s in --projects/-pl resulted in an "
+                    + "empty reactor, please correct %s.", isPlural ? "s" : "", isPlural ? "them" : "it" );
+            throw new MavenExecutionException( message, request.getPom() );
         }
 
-        return result;
+        return excludedProjects;
     }
 
-    private List<MavenProject> includeAlsoMakeTransitively( List<MavenProject> projects, MavenExecutionRequest request,
-                                                            ProjectDependencyGraph graph )
+    private Set<MavenProject> getAlsoMakeDependencies( Set<MavenProject> projectsToActivate,
+                                                       MavenExecutionRequest request, ProjectDependencyGraph graph )
             throws MavenExecutionException
     {
-        List<MavenProject> result = projects;
-
         String makeBehavior = request.getMakeBehavior();
         boolean makeBoth = MavenExecutionRequest.REACTOR_MAKE_BOTH.equals( makeBehavior );
 
         boolean makeUpstream = makeBoth || MavenExecutionRequest.REACTOR_MAKE_UPSTREAM.equals( makeBehavior );
         boolean makeDownstream = makeBoth || MavenExecutionRequest.REACTOR_MAKE_DOWNSTREAM.equals( makeBehavior );
 
-        if ( StringUtils.isNotEmpty( makeBehavior ) && !makeUpstream && !makeDownstream )
+        if ( !makeUpstream && !makeDownstream )
         {
-            throw new MavenExecutionException( "Invalid reactor make behavior: " + makeBehavior,
-                    request.getPom() );
-        }
-
-        if ( makeUpstream || makeDownstream )
-        {
-            Set<MavenProject> projectsSet = new HashSet<>( projects );
-
-            for ( MavenProject project : projects )
+            if ( StringUtils.isNotEmpty( makeBehavior ) )
             {
-                if ( makeUpstream )
-                {
-                    projectsSet.addAll( graph.getUpstreamProjects( project, true ) );
-                }
-                if ( makeDownstream )
-                {
-                    projectsSet.addAll( graph.getDownstreamProjects( project, true ) );
-                }
+                throw new MavenExecutionException( "Invalid reactor make behavior: " + makeBehavior, request.getPom() );
             }
 
-            result = new ArrayList<>( projectsSet );
-
-            // Order the new list in the original order
-            List<MavenProject> sortedProjects = graph.getSortedProjects();
-            result.sort( comparing( sortedProjects::indexOf ) );
+            return Collections.emptySet();
         }
 
-        return result;
+        List<MavenProject> emptyList = Collections.emptyList();
+        return projectsToActivate.stream()
```

Comment:
I _think_ this does the same as the previous code, but to be honest, I'm having a hard time to make sure it does. Why did you introduce a bunch of Streams here? Does it give any benefit over the previous, Collection-based approach?

Reply:
Mostly readability, but I'm okay to revert that if you disagree that it reads better.